doc/developer/design/20240522_mir_representation_types.md

```diff
@@ -0,0 +1,133 @@
+3. Elide noop casts, confirming that we plan the problematic joins better.
+
+PR #2 will have a large diff.
+On the plus side, it should be a largely mechanical change: there is a straightforward way to project a `SqlScalarType` to a `ReprScalarType`, and most uses of types in the optimizer will be parametric.
+
+## Minimal Viable Prototype
+
+The representation types will look like the following:
+
+```rust
+pub enum ReprScalarType {
+  Bool,
+  Int16,
+  Int32,
+  Int64,
+  UInt8, // also includes SqlScalarType::PgLegacyChar
+  UInt16,
+  UInt32, // also includes SqlScalarType::{Oid,RegClass,RegProc,RegType}
+  UInt64,
+  Float32,
+  Float64,
+  Numeric,
+  Date,
+  Time,
+  Timestamp  { precision: Option<TimestampPrecision>, },
+  TimestampTz { precision: Option<TimestampPrecision>, },
+  MzTimestamp,
+  Interval,
+  Bytes,
+  Jsonb,
+  String, // also includes SqlScalarType::{VarChar,Char,PgLegacyName}
+  Uuid,
+  Array(Box<ReprScalarType>), // also includes SqlScalarType::Int2Vector
+  List { element_type: Box<ReprScalarType> }, // also includes SqlScalarType::Record
+  Map { value_type: Box<ReprScalarType> },
+  Range { element_type: Box<ReprScalarType> },
+  MzAclItem,
+  AclItem,
+}
+
+pub struct ReprColumnType {
+    pub scalar_type: ReprScalarType,
+    pub nullable: bool,
+}
+```
+
+We'll need to define a projection from `SqlScalarType` to `ReprScalarType` (see [`ScalarType::is_instance_of()`](https://github.com/MaterializeInc/materialize/blob/e4578164fb28a204b43c58ab8ff6c1d3e3b4427f/src/repr/src/scalar.rs#L947) for the correspondence).
+
+## Alternatives
+
+We considered several alternatives:
+
+- (`eq-type`) Change the MIR typechecker to conflate `varchar` and `text` types (and drop calls to `varchar_to_text`).
+- (`eq-indx`) Change index selection to conflate `varchar` and `text` types (and leave calls to `varchar_to_text`).
+- (`remixvc`) Automatically turn add `varchar_to_text` to all indices on a `varchar`.
+- (`remix++`) Automatically wrap every use of a `varchar` with `varchar_to_text`.
+
+The `eq-type` and `eq-indx` approach came up in early discussions.
+These are fundamentally hacks, with the upside of being quick fixes.
+The downside is the addition of a tricky corner case, some risk (what if keeping the types separate matters somewhere we've forgotten about?), and some technical debt/mess-making.
+
+The `remixvc` and `remix++` approaches address the problem at the user/adapter side.
+Such rewriting feels a bit risky/annoying/confusing, and I'm not sure how it would interact with sources.
+
+## Open questions
+
+How should we report `ReprScalarType`s to the user in `EXPLAIN PLAN`s?
+There's a potential for confusion if we report something as type `string` that the user expected to have type `VARCHAR(18)` or something similar.
+Apparently we already treat `string` as a synonym for `text`... can we somehow choose names that will be unambiguous?
+Say, `mzstring` or `charseq` or `utf8`?
```

Comment:
Seems fine to just report them in whatever representation is convenient for us. `MirScalarExpr`s aren't SQL expressions, and are hard to map to SQL expressions; I don't think we need to feel any obligation for the type information that annotates `MirScalarExpr`s to map directly to SQL types, either.

Maybe just stick an `r` prefix in front of it to make it clear(er) it's a lower-level "repr" type? `rstring`, `rint32`, etc.